### ISSUE TITLE
`protected` alike lint to warn a public metod being used outside of its `Class scope`?

### ISSUE THREAD

**Issue by busslina** [dart-lang/sdk#54473]
Lets say I have this code:

```dart
import 'package:meta/meta.dart';

mixin M {
  
  void start() {
    print('Started -- ${method()}');
  }
  
  String method();
}

class A with M {
  @override
  @protected
  String method()=>'Hey Jude';
}

main() {
  
  final obj = A();
  
  obj.start();
  
  final aa = obj.method();
}
```

I want a `warning` triggered on this line: `final aa = obj.method();`.

Would it be possible?
I have been told Dart only supports two scopes: Global and Library

**Comment by eernstg** [comment]
The call site in your example is in the library where `method` is declared. The [documentation](https://pub.dev/documentation/meta/latest/meta/protected-constant.html) states that this situation should not give rise to any diagnostic messages. Hence, it is working as intended.

You could try to argue that it should be changed. It surely wouldn't be a big thing in terms of implementation. But that's a somewhat long shot, given that this feature is _intended_ to work in this manner. Also, a modified behavior would be a breaking change, so it would have to be supported by a compelling justification.

Here's an attempt to give a motivation for the current behavior:

Dart access control is mainly based on the ability of declared names to be private (the declared name being private if it has an underscore as its first character). Privacy is strictly enforced by the language: `_x` in one library is simply considered to be unrelated to `_x` in another library (for instance, you can't override a private method declaration in a library L1 by a private method declaration in a different library L2, because it's impossible to give them "the same name").

`@protected` is not a language mechanism, it is a tool specific feature. In particular, we have "the analyzer can provide feedback" if a library violates the rules associated with this metadata annotation, as opposed to the much more stringent "it is a compile-time error if ...", which is characteristic of language mechanism related rules.

With this in mind, you could say that `@protected` should be "less strict" than privacy, which would imply that it shouldn't provide feedback on anything which is a usage in library L of a declaration in L.

The argument which is usually given when there is a proposal to introduce a kind of access management mechanism which is more fine-grained than an entire library is that every developer who is writing or maintaining a given declaration in a library L1 is also able to search all of L1 in order to determine whether anything is being done which is a violation of the intended usage. Note that we're searching for a specific name, so we can easily and reliably find all usages of that name (and the only doubt we may encounter is caused by invocations on a receiver of type `dynamic`).

In other words, "you can do whatever you want (with respect to access restrictions) inside the same library because you can also fix whatever you don't want".

Note that the same approach has been used with class modifiers: You can declare a class `C` to be `final`, and this means that it is an error to have, among other things, `extends C` or `implements C`. However, this only occurs if `C` and the subtype is declared in different libraries (e.g., you can have `final class C {}` and `base class D extends C {}` as long as both classes are declared in the same library).

This means that "you can do whatever you want as long as it is in the same library" is an informal version of several rules in Dart, not just the ones about privacy. This serves as an informal consistency argument.

**Comment by busslina** [comment]
I explain why I'm asking for this feature (I don't want to change the `protected` behaviour).

I have a class `A` that mixes with the `M` mixin. In `M` I declare an abstract public `method`. The only reason of `method` being public is to be override but it's intended to only be executed privately in `M` (or in one of the sub/classes that mixes `M`).

It would be very helpful to be able to get an advice on this, because in the future I can forget about the `M` implementation so I can see a public method and use it, which is what I want to prevent.

In my use case, the `A` class is private and coexists with an `extension` which is the place where I can call `method` but I want not to.

I appreciate your excellent explanation @eernstg.

**Comment by bwilkerson** [comment]
Have you considered using [`visibleForOverriding`](https://github.com/dart-lang/sdk/blob/main/pkg/meta/lib/meta.dart#L372)? It sounds like it might be what you're looking for.

**Comment by busslina** [comment]
> Have you considered using [visibleForOverriding](https://github.com/dart-lang/sdk/blob/main/pkg/meta/lib/meta.dart#L372)? It sounds like it might be what you're looking for.

Didn't knew about that, but it doesn't trigger the warning on a different class/extension placed on the same `library`. (Like `protected`).

But it scope is


```
Used to annotate an instance member that was made public so that it could be
overridden but that is not intended to be referenced from outside the
defining library.
```

**Comment by eernstg** [comment]
Thanks for the extra information, @busslina!

We did have some discussions about a particular kind of access restriction that might fit your needs (not sure, but it might be relevant):

We could introduce a restriction on an instance member specifying that it can only be invoked with `this` as the receiver. This would make `obj.method()` in `main` an error because the receiver is `obj`, but it would allow the following:

```dart
mixin M {
  void start() {
    print('Started -- ${this.method()}'); // OK.
  }

  // Something like `@receiverMustBeThis`.
  String method();
}

class A with M {
  void doCall() => print(method()); // OK, receiver is implicitly `this`.

  @override
  String method() => 'Hey Jude';
}

main() {
  final obj = A();
  obj.start();
  obj.doCall();
}
```

If this is done as a metadata annotation plus some analyzer checks then it would serve as a software engineering device, a soft access restriction, nudging developers in the direction of calling that method only on `this`. (This could be controversial if it is stronger than the language level privacy, that is, if it is enforced also in the declaring library.)

However, it could also be a language mechanism, in which case it would be enforced strictly, and it could be used to provide firm guarantees that private methods exist: if the receiver is `this`, and the interface of `this` has a private method `_m` then `this` can in some situations be guaranteed to have an implementation of `_m`.

**Comment by busslina** [comment]
I like this idea!

@eernstg
